cpmpy/expressions/globalconstraints.py

```diff
@@ -272,6 +273,105 @@ def value(self):
+        Nodes which aren't part of the subcircuit, should self loop i.e. x[i] = i.
+        The size of the subcircuit should be strictly greater than 1, so not all stops can selfloop.
+        When a startIndex is provided, it will be treated as the start of the subcircuit and will be garuanteed
+        to be inside the subcircuit.
+    """
+
+    def __init__(self, *args, startIndex:int=None):
+        flatargs = flatlist(args)
+        # Ensure all args are integer successor values
+        if any(is_boolexpr(arg) for arg in flatargs):
+            raise TypeError("SubCircuit global constraint only takes arithmetic arguments: {}".format(flatargs))
+        # Ensure startIndex is an integer
+        if not isinstance(startIndex, int) and startIndex is not None:
+            raise TypeError("SubCircuit global constraint's startIndex argument must be an integer: {}".format(startIndex))
+        super().__init__("subcircuit", flatargs)
+        self.startIndex = startIndex
+        # Ensure there are at least two stops to create a circuit with
+        if len(flatargs) < 2:
+            raise CPMpyException('SubCircuit constraint must be given a minimum of 2 variables')
+
+    def decompose(self):
+        """
+            Decomposition for SubCircuit
+
+            A mix of the above circuit implementation, with elements from the Minizinc implementation for the support of optional visits:
+            https://github.com/MiniZinc/minizinc-old/blob/master/lib/minizinc/std/subcircuit.mzn
+        """
+
+        from .python_builtins import all as cpm_all
+        
+        startIndex = self.startIndex
+        succ = cpm_array(self.args) # Successor variables
+        n = len(succ)
+        order = intvar(0, n-1, shape=n) # Order variables of the stops within the subcircuit (if part of it, otherwise a remaining value in the domain to satisfy AllDiff).
+        instantiated = boolvar(shape=n) # Whether a stop is part of the subcircuit.
+        endIndex = intvar(0, n-1) # Index of the last stop in the subcircuit, before looping back to the startIndex.
+         # If a startIndex is not supplied, create an additional auxilary variable
+        if startIndex is None: startIndex = intvar(0, n-1)
+
+        constraining = []
+        constraining += [AllDifferent(succ)] # All stops should have a unique successor.
+        constraining += [AllDifferent(order)] # All stops should have a unique order.
+        constraining += list( instantiated.implies(succ < len(succ)) ) # Successor values should remain within domain.
+
+        defining = []
+        defining += [order[startIndex] == 0] # The ordering starts at the startIndex stop.   
+        defining += [cpm_all(((instantiated[i] == True), (i != endIndex))).implies(order[succ[i]] == (order[i] + 1)) for i in range(0,n)] # If a stop is on the subcircuit and it is not the last one, than its successor should have +1 as order.
```

Comment:
This is a bit long for a list comprehension ;)

Can you make this a normal for-loop instead? Also, use `&` instead of wrapping two expressions into an "All" this makes it more readable imho